src/v/iceberg/partition.cc

```diff
@@ -0,0 +1,61 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "iceberg/partition.h"
+
+namespace iceberg {
+
+std::optional<partition_spec> partition_spec::resolve(
+  const unresolved_partition_spec& spec, const struct_type& schema_type) {
+    auto cur_field_id = partition_field::id_t{1000};
+    chunked_vector<partition_field> fields;
+    for (const auto& field : spec.fields) {
+        const auto* cur_struct_type = &schema_type;
+        std::optional<nested_field::id_t> field_id;
+        for (const auto& s : field.source_name) {
+            if (!cur_struct_type) {
+                return std::nullopt;
+            }
+
+            const nested_field* field = nullptr;
+            for (const auto& f : cur_struct_type->fields) {
+                if (f->name == s) {
+                    field = f.get();
+                    break;
+                }
+            }
+            if (!field) {
+                return std::nullopt;
+            }
+
+            field_id = field->id;
+            cur_struct_type = std::get_if<struct_type>(&field->type);
+        }
+
+        if (!field_id) {
+            // empty source field name
+            return std::nullopt;
+        }
```

Comment:
nit: maybe worth pulling this inner loop into some `find_field_by_name(const struct_type&, const vector<sstring>& target_field)`? The deep nesting is otherwise a little tricky to read through